libs/base/Data/List.idr

```diff
@@ -358,19 +358,31 @@ export
 zip3 : List a -> List b -> List c -> List (a, b, c)
 zip3 = zipWith3 \x, y, z => (x, y, z)
 
+||| Split a list by applying a function into two lists
+export
+unzipWith : (a -> (b, c)) -> List a -> (List b, List c)
+unzipWith f [] = ([], [])
+unzipWith f (x :: xs) with (unzipWith f xs)
+  unzipWith f (x :: xs) | (lefts, rights) = (fst (f x) :: lefts, snd (f x) :: rights)
```

Comment:
I don't think we need to use a `with` (that means generating an auxiliary definition)
and we should probably avoid duplicating the `f x` call. So I'd write something like
this instead (and something similar for `unzipWith3`):

```suggestion
unzipWith f (x :: xs)
  = let (b, c) = f x
        (bs, cs) = unzipWith f xs
    in (b :: bs, c :: cs)
```